README.md
# eea.searchserver.js
Node.js search server library for EEA search apps

To be used with an express Node.js application

### Features
* Proxy for any Elastic backend (so Elastic doesn't have to be exposed
  publicly)
* EEA Template Invalidation Routines
  * Automatic template loading if templates do not exist locally
  * On demand template loading via routes or util function
* API for Elastic management commands
* Framework for easily adding index management commands
* Entrypoint for any of EEA's Elastic Search Docker applications

### Install

```bash
npm install git://github.com/eea/eea.searchserver.js
```

## Contents

### Server
#### Initial setup
Start by writing the base configurtion file:

```js
{
  "http": {
    "port": 8080         // Port on which the app will listen, default(3000)
  },
  "elastic": {           // Remote Elastic Endpoint configuration
    "host":
      "my-elastic-host", // Host running an Elastic Server (required)
    "path":
      "/elastic/",       // Path on host to the Elastic Server (default /)
    "port": 80,          // Port on which Elastic listens (default 9200)
    "index":
      "data",            // Index to be queried (required)
    "type":
      "resource"         // Type to be queried (required)
  },
  "external_templates": { // External template service (optional)
    "local_path":
      "/path/to/external_templates", // Path to save external templates locally
    "host":
      "www.eea.europa.eu",          // Host to query for external templates
    "head_path":                    // Path on host for the page HEAD
      "/templates/v2/getRequiredHead?jsdisable=all",
    "header_path":                  // Path on host for the page HEADER
      "/templates/v2/getHeader?jsdisable=all",
    "footer_path":                  // Path on host for the page FOOTER
      "/templates/v2/getFooter"
  },
  "switch_condition_value_percent": -5  // resync protection condition
}
```

#### Running a simple express app

```js
var eeasearch = require('eea-searchserver');
// create a base app
var app = express();
...
// load app settings
var settingsFile = '/path/to/settings.json';
// create server
server = eeasearch.Server(app, settingsFile)
// start the app
server.run('runserver', [], function(err, server) {
    console.log("Started server");
}
```

#### Easy to add management commands:

```js
var eeasearch = require('eea-searchserver');
// create a base app
var app = express();
...
// add management commands
app.set('managementCommands', {'cmd': function(args) { console.log(args); }});
// load app settings
var settingsFile = '/path/to/settings.json';
// create server
server = eeasearch.Server(app, settingsFile, function(err, server) {
    if (err) console.log("The app was poorly configured: " + err.message);
})
// start the app
server.run('cmd', ['foo', 'bar'], function(err, server) {
    if (err) {
        console.log("Something went wrong when running the command");
    }
    console.log("Ran custom command, should see ['foo', 'bar'] on screen");
}
```

### Routes
#### Invalidate Templates

Any POST request to ```invalidate_templates``` will get the templates from
the host which was set up in the ```settings.json``` file and save them to the
local path.

```js
var app = express();
app.post('/invalidate_templates', eeasearch.routes.invalidateTemplates);
```

#### Update external configurations

Any request to ```/tools/update_external_configs``` will get the latest changes
from the host which was set up in the ```settings.json``` file and save them in
elasticsearch as a cache document with _id ```externals```.

#### Elastic Proxy

Any GET request with the parameter ```source``` set as a valid Elastic query
or Any POST request with the body an valid Elastic query will be forwarded
to the Elastic host set up in the ```settings.json``` file.

The queries will be carried out to the index and type configured in
```settings.json``` .

```js
var app = express();
app.post('/api', eeasearch.routes.elasticProxy);
```

### Middleware
#### Automatic template loading

Use this to load templates automatically on the first request on the app.
This will ensure that the external templates exist on disk before
rendering the page.

```js
var app=express();
app.use(eeasearch.middleware.templateRequired);
```

### Util
#### On demand template loading

This will send requests and update the external templates using the
configuration in ```settings.json```

```js
eeasearch.util.invalidateTemplates()
```


### esAPI

A lightweight REST api for Elastic index management commands.
To chain the commands use chained calls like in the following example.
Otherwise, the order of the requests is asynchronous.

The REST calls are executed when ```.execute()``` is called. If any error
occurs while executing a request, the others are not run.

__Note:__. This API does not use ```settings.json``` configured Elastic host
__Note:__. A successful run means any response from the server (500, 200, etc.)

```js
var options = {
    "es_host": "http://host:port/path/to/endpoint",
    "auth": {"user": user, "pass": pass} # optional if elastic needs auth
}
esAPI(options).DELETE("/index/type/_mappings",
                      function(err, code, header, body) {
                        // Code can be any valid HTTP code
                        // HTTP error codes (500, 403, ...)
                        // are not marked as errors
                        CallbackLogic();
                      })
              .PUT("/index/type/_mappings",
                      function(err, code, header, body) {
                        OtherLogic();
                      })
              .execute();
```


CHANGES.md
# Changes
For previous changes checkout the corresponding git tag commits

### 1.3.17

 * Add polyfill for nodeList forEach as IE 11 doesn't support it
   natively 
   [ichim-david refs #90612]
### 1.3.16

 * Strip html entities when parsing elastic search results
   [ichim-david refs #90612]

lib/esAPI.js
#!/usr/bin/nodejs
/* eslint-disable no-console */

var http = require('http');
var parse = require('csv-parse/lib/sync');
var nconf = require('nconf');

var sendRequest = function(method, url, auth, body, callback) {
    if (!url.indexOf('http://') == 0)
        url = 'http://' + url;

    var urlparse = require('url');
    var parsedUrl = urlparse.parse(url);

    var options = {
        host: parsedUrl.hostname,
        path: parsedUrl.path,
        method: method,
        auth: auth,
        port: parsedUrl.port ? parsedUrl.port : 80
    };

console.log(options);
    var req = http.request(options, function(res) {
    console.log("XXXXXXXXXXXXXX");
        res.setEncoding('utf8');
        var body = '';
        res.on('data', function (chunk) {
            body += chunk;
        });
        res.on('end', function() {
            callback(null, res.statusCode, res.header, body);
        });
        res.on('error', function() {
            console.log("XXXXXXXXXXXXXX");
        });
    });
//if ((method === 'PUT') || (method === 'POST')){
    req.setHeader('Content-Type', 'application/json');
//}
    req.on('error', function(err) { console.log("XXXXXXXXXXXXXX");callback(err); });
    req.write(body);
    req.end();
};

var total_queries = 0;
var success_calls = 0;
if (nconf.get().hasOwnProperty('switch_condition_value_percent')) {
    var SWITCH_CONDITION_VALUE_PERCENT = nconf.get()['switch_condition_value_percent'];
} else {
    var SWITCH_CONDITION_VALUE_PERCENT = -5;
}
var default_callback = function(text, showBody) {
    if (showBody === undefined){
        showBody = true;
    }
    return function(err, statusCode, header, body) {
        console.log(text);
        if (err) {
            console.log(err.message);
        } else {
            console.log('  Successfuly ran query');
            console.log('  ResponseCode: ' + statusCode);
            if (showBody === true){
                console.log('  ' + body);
            }
        }
    };
};

var switch_alias_callback = function(text, elastic, esoptions, showBody) {
    if (showBody === undefined){
        showBody = true;
    }
    return function(err, statusCode, header, body) {
        try{
            console.log(text);
            if (err) {
                console.log(err.message);
            } else {
                console.log('  Successfuly ran query');
                console.log('  ResponseCode: ' + statusCode);
                if (showBody === true){
                    console.log('  ' + body);
                }
            }
            success_calls = success_calls + 1;
        }
        catch(e){
            console.log("\n\n ERROR: " + e + "\n\n" );
        }
        finally{
            if (success_calls >= total_queries){
                applyAlias(elastic, esoptions);
            }
        }
    };
};

function fetchQuery(query, endpoint, elastic, esoptions, analyzers, fetchcallback) {
    var properties = analyzers.mappings[elastic.type].properties;
    var analyzers = analyzers.settings.analysis.analyzer;
    var batch_head = '{"index":{}}';

    var SparqlClient = require('sparql-client');
    var client = new SparqlClient(endpoint);
    client.query(query).execute(function(error, results){
        var rows_str = "";
        for (var i = 0; i < results.results.bindings.length; i++){
            var toindex = {};
            for (var j = 0; j < results.head.vars.length; j++){
                if (results.results.bindings[i][results.head.vars[j]] !== undefined){
                    if (results.head.vars[j] !== "_id"){
                    toindex[results.head.vars[j]] = results.results.bindings[i][results.head.vars[j]].value;
                    if (elastic.enableValuesCounting){
                        var items_count = 1;
                        var field_name = results.head.vars[j];
                        if ((properties[field_name] !== undefined) &&
                            (properties[field_name].analyzer !== undefined) &&
                            (analyzers[properties[field_name].analyzer].type === 'pattern')){
                            var pattern = analyzers[properties[field_name].analyzer].pattern;
                            var values = parse(results.results.bindings[i][field_name].value, pattern.trim());
                            if (values[0]){
                                items_count = values[0].length;
                            }
                        }
                        toindex["items_count_" + results.head.vars[j]] = items_count;
                    }
                    }
                }
            }
            rows_str += batch_head.split("{}").join('{"_id":"' + results.results.bindings[i]["_id"].value + '"}');
            rows_str += "\n";
            rows_str += JSON.stringify(toindex);
            rows_str += "\n";
        }
        new esAPI(esoptions)
            .POST(elastic.real_index+"/"+elastic.type+"/_bulk", rows_str, 
                  switch_alias_callback("indexed " + results.results.bindings.length + " rows", elastic, esoptions, false))
            .execute();

        fetchcallback(results.results.bindings.length);
    });
}

var bigTotal = 0;
function fetchQueries(queries, endpoint, elastic, esoptions, analyzers, idx){
    if (idx === undefined){
        idx = 0;
    }

    if (idx === queries.length){
        return;
    }
    console.log("Fetching query #", idx + 1, " of ", queries.length);
    if (queries[idx].filters){
        console.log("Filters:");
        console.log(queries[idx].filters);
    }
    fetchQuery(queries[idx].query, endpoint, elastic, esoptions, analyzers, function(indexed){
        bigTotal += indexed;
        console.log("Total indexed:", bigTotal);
        fetchQueries(queries, endpoint, elastic, esoptions, analyzers, idx + 1);
    });
}

function getBlueIndex(){
    var elastic = nconf.get()['elastic'];
    elastic.real_index = elastic.index + "_blue";
    return elastic;
}
function getGreenIndex(){
    var elastic = nconf.get()['elastic'];
    elastic.real_index = elastic.index + "_green";
    return elastic;
}
function applyAlias(elastic, esoptions){
    var request = require('sync-request');
    var elastic_blue = getBlueIndex();
    var elastic_green = getGreenIndex();
    var current_elastic;
    var next_elastic;

console.log(esoptions);
    if (elastic_green === undefined){
        console.log("SKIP ALIAS, the second index is not present!");
        return;
    }

    if(elastic_blue.real_index == elastic.real_index){
        /* current index is green */
        current_elastic = elastic_green;
        next_elastic = elastic_blue;
    } else {
        /* current index is blue */
        current_elastic = elastic_blue;
        next_elastic = elastic_green;
    }
    console.log('\nATTEMPT OF SWITCHING ALIAS');
    console.log('Current alias is on ' + current_elastic.real_index);

    var indexed_current_url = esoptions.es_host+'/'+
                              current_elastic.real_index+'/_count';
    indexed_current_url = 'http://' + indexed_current_url.split("//").join("/");
    var indexed_next_url = esoptions.es_host+'/'+
                           next_elastic.real_index+'/_count';
    indexed_next_url = 'http://' + indexed_next_url.split("//").join("/");
    console.log(indexed_current_url);
    console.log(indexed_next_url);
    var count_current = 0;
    try{
        var res = request('GET', indexed_current_url);
        var count_current = JSON.parse(res.getBody('utf8')).count;
        if (isNaN(count_current)){
            count_current = 1;
        }
    }catch(e){
        count_current = 1;
    }
    var count_next = 0;
    try{
console.log("try");
        var res = request('GET', indexed_next_url);
        console.log(JSON.parse(res.getBody('utf8')))
        count_next = JSON.parse(res.getBody('utf8')).count;
        if (isNaN(count_next)){
            count_next = 0;
        }
    }catch(e){
        count_next = 0;
    }
    console.log(count_current);
    console.log(count_next);
    var increment_percent = ((count_next-count_current)/count_current)*100;
    // if new index size is > then old index size of the SWITCH_CONDITION_VALUE_PERCENT
    // switch alias, none otherwise
    if (increment_percent > SWITCH_CONDITION_VALUE_PERCENT){
        var remove_alias = '{"actions":[{"remove":{"alias":"'+current_elastic.index+'","index":"'+current_elastic.real_index+'"}}]}';
        var add_alias = '{"actions":[{"add":{"alias":"'+next_elastic.index+'","index":"'+next_elastic.real_index+'"}}]}';
        new esAPI(esoptions)
            .PUT("/_aliases", remove_alias, default_callback(
                "REMOVE OLD ALIAS: " + current_elastic.real_index))
            .PUT("/_aliases", add_alias, default_callback(
                "SUCCESS SWITCH ALIAS: Alias is now on " + next_elastic.real_index))
            .execute();
    } else {
        console.log('ABORT SWITCH ALIAS: The new index size is lower then old');
    }
}

var esAPI = function(options) {
    this.callbacks = [];
    this.operations = [];
    this.options = options;
};

esAPI.prototype = {
    PUT: function(endpoint, body, callback) {
        var idx = this.callbacks.length;
        var callbacks = this.callbacks;
        var options = this.options;

        var putImpl = function() {
            sendRequest('PUT',
                options.es_host + endpoint,
                options.auth,
                JSON.stringify(body),
                callbacks[idx]);
        };

        this.callbacks.push(callback);
        this.operations.push(putImpl);
        return this;
    },

    POST: function(endpoint, body, callback) {
        var idx = this.callbacks.length;
        var callbacks = this.callbacks;
        var options = this.options;

        if (typeof(body) === 'object'){
            body = JSON.stringify(body);
        }

        var putImpl = function() {
            sendRequest('POST',
                options.es_host + endpoint,
                options.auth,
                body,
                callbacks[idx]);
        };

        this.callbacks.push(callback);
        this.operations.push(putImpl);
        return this;
    },

    DELETE: function(endpoint, callback) {
        var idx = this.callbacks.length;
        var callbacks = this.callbacks;
        var options = this.options;

        var delImpl = function() {
            sendRequest('DELETE',
            options.es_host + endpoint,
            options.auth, '',
            callbacks[idx]);
        };

        this.callbacks.push(callback);
        this.operations.push(delImpl);
        return this;
    },

    execute: function() {
        var callbacks = this.callbacks;
        var operations = this.operations;

        for (var i = this.operations.length - 2; i >= 0; i--) {
            // forcing block-level scoping
            var f = function(idx) {
                var oldCb = callbacks[idx];
                var next = operations[idx + 1];
                callbacks[i] = function(err, s, e, b) {
                    if (err) {
                        oldCb(err);
                    } else {
                        oldCb(null, s, e, b);
                        next();
                    }
                };
            };
            f(i);
        }
        this.operations[0]();
    },

    createQueries: function(endpoint, base_query, filters_query, esoptions, callback){
        var queries = [];
        if (!filters_query){
            queries.push({query:base_query, filters:[]});
            callback(queries,esoptions);
            return;
        }
        var SparqlClient = require('sparql-client');
        var client = new SparqlClient(endpoint);
        client.query(filters_query).execute(function(error, results){
            for (var i = 0; i < results.results.bindings.length; i++){
                var tmp_query = {query:base_query, filters:{}};
                for (var j = 0; j < results.head.vars.length; j++){
                    var attr = results.head.vars[j];
                    var value = results.results.bindings[i][results.head.vars[j]].value;
                    tmp_query.query = tmp_query.query.split("<" + attr + ">").join(value);
                    tmp_query.filters[attr] = value;
                }
                queries.push(tmp_query);
                total_queries = queries.length;
            }
            callback(queries, esoptions);
        });
    },

    indexFromQuery: function(endpoint, queryTemplate, filtersQuery, elastic, analyzers) {
        if (elastic.real_index === undefined){
            elastic.real_index = elastic.index;
        }
        this.createQueries(endpoint, queryTemplate, filtersQuery, this.options, function(queries, esoptions){
            new esAPI(esoptions)
                .DELETE(elastic.real_index, default_callback('Deleting index (if it exists)'))
                .PUT(elastic.real_index, analyzers,
                    function(){
                        fetchQueries(queries, endpoint, elastic, esoptions, analyzers);
                    })
                .execute();
        });
    },

    syncFromQuery: function(endpoint, queryTemplate, filtersQuery, analyzers) {
        var request = require('request');
        var elastic_blue = getBlueIndex();
        var elastic_green = getGreenIndex();
        var next_elastic = elastic_blue;
        var parent = this;
        /* get current index by alias */
        request(
            'http://' + parent.options.es_host + '_aliases',
            function (error, response, body) {
                if (!error && response.statusCode == 200) {
                    var data = JSON.parse(body);

                    if(data.hasOwnProperty(elastic_blue.real_index) && data[elastic_blue.real_index]['aliases'].hasOwnProperty(elastic_blue.index)){
                        /* current index is blue */
                        next_elastic = elastic_green;
                    }
                    else if(data.hasOwnProperty(elastic_green.real_index) && data[elastic_green.real_index]['aliases'].hasOwnProperty(elastic_green.index)){
                        /* current index is green */
                        next_elastic = elastic_blue;
                    }
                    /* create the new index */
                    parent.indexFromQuery(endpoint, queryTemplate, filtersQuery, next_elastic, analyzers);
                } else {
                    console.log(error);
                }
            }
        );
    }
};

module.exports = esAPI;
